### PR TITLE
Use task configuration avoidance #141

### PR DESCRIPTION
**Pull request by f4lco**
At first some easy instances.

I do believe #142 started with the hardest part of all. @boris-petrov your suggested fix does not work 😬   Do you remember any backstory regarding `integrationTestTaskAssigned`? I surely have problems to understand what's going on there.

### CODE REVIEW

**Comment by f4lco** [comment]
That's the current state of the PR:

| Metric                                | Before | After | Delta |
| ------------------------------------- | ------ | ----- | ----- |
| Total configuration time              | 15s    | 8s    | -7s   |
| Total tasks                           | 4048   | 4048  | +0    |
| Created immediately                   | 2236   | 526   | -1710 |
| Created during configuration          | 263    | 461   | +198  |
| Created during task execution         | 2      | 2     | +0    |
| Created during task graph calculation | 25     | 25    | +0    |
| Not created                           | 1522   | 3034  | +1512 |

File: libs/gretty/src/main/groovy/org/akhikhl/gretty/GrettyPlugin.groovy [hunk 1]
```diff
@@ -280,17 +280,20 @@ class GrettyPlugin implements Plugin<Project> {
...
 
-      project.task('prepareInplaceWebAppClasses', group: 'gretty') {
+      project.tasks.register('prepareInplaceWebAppClasses') {
+        group = 'gretty'
         description = 'Compiles classes of this web-app and all overlays (if any)'
-        dependsOn project.tasks.classes
+        dependsOn project.tasks.named('classes')
+        // FIXME does not use task configuration avoidance #41
         for(String overlay in project.gretty.overlays)
-          dependsOn "$overlay:prepareInplaceWebAppClasses"
```
Comment (boris-petrov) [inline libs/gretty/src/main/groovy/org/akhikhl/gretty/GrettyPlugin.groovy:287]: Whitespace issue here. Not sure why it compiles without the space?! :smile:

File: libs/gretty/src/main/groovy/org/akhikhl/gretty/GrettyPlugin.groovy [hunk 2]
```diff
@@ -263,11 +263,11 @@ class GrettyPlugin implements Plugin<Project> {
         into "${project.buildDir}/inplaceWebapp"
       }
 
-      project.task('createInplaceWebAppFolder', group: 'gretty') {
-
+      project.tasks.register('createInplaceWebAppFolder') {
+        group = 'gretty'
         description = 'Creates directory ${buildDir}/inplaceWebapp'
 
-        dependsOn project.tasks.prepareInplaceWebAppFolder
```
Comment (boris-petrov) [inline libs/gretty/src/main/groovy/org/akhikhl/gretty/GrettyPlugin.groovy:270]: When in a `project.tasks.register` block, there is no need to use `tasks.named` inside. The block is "lazy" anyways so this change (and all the similar ones in this PR) could be reverted.

See [here](https://docs.gradle.org/current/userguide/task_configuration_avoidance.html#sec:task_configuration_avoidance_pitfalls).

**Comment by boris-petrov** [comment]
@f4lco - thanks! I added a couple of comments, please check them out. Otherwise the numbers look amazing! 1512 tasks not created? That's really good!

**Comment by boris-petrov** [comment]
As for #142 - my suggested "fix" there doesn't fix the tests, as I've noted, but is the correct mapping from the old code to the new one. Otherwise the current state without my requested change does something else and hits a Gretty warning.

File: libs/gretty/src/main/groovy/org/akhikhl/gretty/GrettyPlugin.groovy (hunk 1 above)
Comment (f4lco) [reply libs/gretty/src/main/groovy/org/akhikhl/gretty/GrettyPlugin.groovy:287]: Groovy 🤷‍♂️

File: libs/gretty/src/main/groovy/org/akhikhl/gretty/GrettyPlugin.groovy (hunk 2 above)
Comment (f4lco) [reply libs/gretty/src/main/groovy/org/akhikhl/gretty/GrettyPlugin.groovy:270]: OK will do.

**Comment by boris-petrov** [comment]
Looks great! Can you please check if the numbers match the ones from [yesterday](https://github.com/gretty-gradle-plugin/gretty/pull/203#issuecomment-796565445)?

**Comment by f4lco** [comment]
@boris-petrov checked, indeed the number match. No regression.

**Comment by boris-petrov** [comment]
Great, thanks a lot! In that case it's ready for merging? I'm happy with it so go ahead when you feel like it. :)

**Comment by f4lco** [comment]
OK thanks let's just wait for the build.

I think it will be already an improvement to merge as-is. Eventually I'll try a backport on 3.x.
But to be honest I don't know when and how to tackle the hard parts. They look incompatible with configuration avoidance at first glance. And then there is the integration test plugin, too. I have a hard time understanding why it was written in this very particular way and how to upgrade it.

**Comment by boris-petrov** [comment]
The backport should be as simple as a cherry-pick, I don't think there will be conflicts.

As for the "hard parts" - well, we can leave them for now. I'm also struggling to understand 90% of Gretty and the way it is written but yeah. :D As for the integration test plugin - this is the one that users of Gretty use to write their own integration tests or the one that Gretty uses internally for its own tests? Or are they the same? In any case we can leave this for last - the "main" plugin is more important.

As you said, this is a good starting point here. We'll do more when we can. Thanks again!

**Comment by boris-petrov** [comment]
@f4lco  - I cherry-picked it in `gretty-3.x`.

**Comment by f4lco** [comment]
Thanks @boris-petrov and this is the rundown of numbers for 3.x:

| Metric                                | Before | After | Delta |
| ------------------------------------- | ------ | ----- | ----- |
| Total configuration time              | 25s    | 15s   | -10s   |
| Total tasks                           | 5220   | 5220  | +0    |
| Created immediately                   | 3005   | 787   | -2218 |
| Created during configuration          | 380    | 824   | +444  |
| Created during task execution         | 2      | 2     | +0    |
| Created during task graph calculation | 25     | 25    | +0    |
| Not created                           | 1808   | 3582  | +1774 |

**Pull request merged**